### PR TITLE
Retry couple times if restoring a backup failed

### PR DESCRIPTION
**Pull request by mantissahz**
simple retry to give another change for restoring to get the lock.

longhorn/longhorn#3055

Signed-off-by: James Lu <james.lu@suse.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/gorilla/handlers v1.4.2
 	github.com/gorilla/mux v1.7.3
 	github.com/gorilla/websocket v1.4.2 // indirect
+	github.com/jpillora/backoff v1.0.0
```
Comment (shuo-wu) [inline go.mod:15]: Oh, it's a new backoff pkg. We are using `"k8s.io/client-go/util/flowcontrol"` in longhorn-manager repo. Not sure which one is better for this case

File: pkg/backup/main.go [hunk 2]
```diff
@@ -136,14 +143,33 @@ func DoBackupRestore(backupURL string, toFile string, restoreObj *replica.Restor
...
+
 	config := &backupstore.DeltaRestoreConfig{
 		BackupURL: backupURL,
 		DeltaOps:  restoreObj,
 		Filename:  toFile,
 	}
 
-	if err := backupstore.RestoreDeltaBlockBackup(config); err != nil {
-		return err
+	for {
```
Comment (shuo-wu) [inline pkg/backup/main.go:160]: I don't think we can add a wait & retry loop here. This will block the restore call then lead to timeout errors.

**Review by shuo-wu — CHANGES_REQUESTED** [review]
Right now the restore handling in the sync agent server consists of 2 parts:
1. Sync part: Init the restore progress then return immediately. This cannot be blocked.
2. Async part: wait for the restore complete then do some wind-up operations

File: go.mod (hunk 1 above)
Comment (mantissahz) [reply go.mod:15]: I think `"k8s.io/client-go/util/flowcontrol"` is good enough, I would turn to use this pkg.

File: pkg/backup/main.go (hunk 2 above)
Comment (mantissahz) [reply pkg/backup/main.go:160]: Yes, so It has just 5 times to retry and timeout in one minute.
I should turn to `Async part` to handle the lock acquisition failed, right?

Comment (mantissahz) [reply pkg/backup/main.go:160]: or Could I check the getting lock failed here [sync() and refresh()](https://github.com/longhorn/longhorn-manager/blob/master/controller/engine_controller.go#L727) and retry?

**Comment by mantissahz** [comment]
Close this PR, and retry the restore in upper layer.

**Pull request closed without merging**